include/wil/coroutine.h

```diff
@@ -0,0 +1,810 @@
+#include <wil/wistd_memory.h>
+#include <wil/wistd_type_traits.h>
+#include <wil/result_macros.h>
+
+namespace wil
+{
+    // There are three general categories of T that you can
+    // use with a task. We give them these names:
+    //
+    // T = void ("void category")
+    // T = some kind of reference ("reference category")
+    // T = non-void non-reference ("object category")
+    //
+    // Take care that the implementation supports all three categories.
+    //
+    // There is a sub-category of object category for move-only types.
+    // We designed our task to be co_awaitable only once, so that
+    // it can contain a move-only type. Any transfer of T as an
+    // object category must be done as an rvalue reference.
+    template<typename T>
+    struct task;
+
+    template<typename T>
+    struct com_task;
+}
+
+namespace wil::details::coro
+{
+    template<typename T>
+    struct task_promise;
+
+    // Unions may not contain references, C++/CX types, or void.
+    // To work around that, we put everything inside a result_wrapper
+    // struct, and put the struct in the union. For void,
+    // we create a special empty structure.
+    //
+    // get_value returns rvalue reference to T for object
+    // category, or just T itself for void and reference
+    // category.
+    //
+    // We take advantage of the reference collapsing rules
+    // so that T&& = T if T is reference category.
+
+    template<typename T>
+    struct result_wrapper
+    {
+        T value;
+        T get_value() { return wistd::forward<T>(value); }
+    };
+
+    template<>
+    struct result_wrapper<void>
+    {
+        void get_value() { }
+    };
+
+
+    // The result_holder is basically a
+    // std::variant<std::monotype, T, std::exception_ptr>
+    // but with these extra quirks:
+    // * The only valid transition is monotype -> something-else.
+    //   Consequently, it does not have valueless_by_exception.
+
+    template<typename T>
+    struct result_holder
+    {
+        // The content of the result_holder
+        // depends on the result_status:
+        //
+        // empty: No active member.
+        // value: Active member is wrap.
+        // error: Active member is error.
+        enum class result_status { empty, value, error };
+
+        result_status status{ result_status::empty };
+        union variant
+        {
+            variant() {}
+            ~variant() {}
+            result_wrapper<T> wrap;
+            std::exception_ptr error;
+        } result;
+
+        // set_value will be called with
+        //
+        // * no parameters (void category)
+        // * The reference type T (reference category)
+        // * Some kind of reference to T (object category)
+        //
+        // Set the status after constructing the object.
+        // That way, if object construction throws an exception,
+        // the holder remains empty.
+        template<typename...Args>
+        void set_value(Args&&... args)

```

Comment:
Consider renaming these to something like `emplace` or `emplace_value` for consistency with the language